mathesar/api/db/viewsets/records.py

```diff
@@ -61,23 +61,32 @@ def list(self, request, table_pk=None):
                 status_code=status.HTTP_400_BAD_REQUEST
             )
 
-        serializer = RecordSerializer(records, many=True)
+        # TODO: Prefetch column names to avoid N+1 queries
+        columns_map = get_column_id_name_bidirectional_map(table_pk)
+        serializer = RecordSerializer(
+            records,
+            many=True,
+            context={'request': request, 'columns_map': columns_map.inverse, 'table_pk': table_pk}
+        )
         return paginator.get_paginated_response(serializer.data)
 
     def retrieve(self, request, pk=None, table_pk=None):
         table = get_table_or_404(table_pk)
         record = table.get_record(pk)
         if not record:
             raise NotFound
-        serializer = RecordSerializer(record)
+        columns_map = get_column_id_name_bidirectional_map(table.id)
+        serializer = RecordSerializer(record, context={'columns_map': columns_map.inverse, 'table_pk': table_pk})
         return Response(serializer.data)
 
     def create(self, request, table_pk=None):
         table = get_table_or_404(table_pk)
         # We only support adding a single record through the API.
         assert isinstance((request.data), dict)
+        columns_map = get_column_id_name_bidirectional_map(table.id)
+        data = {columns_map[int(column_id)]: value for column_id, value in request.data.items()}
```

Comment:
Deserialization should be handled in the `.to_internal_value()` method of the serializer rather than in the viewset here. Also applies to other instances.